source/Calamari.AzureResourceGroup/DeployAzureResourceGroupBehaviour.cs

```diff
@@ -32,197 +41,139 @@ public DeployAzureResourceGroupBehaviour(TemplateService templateService, IResou
-                    if (deployment.Properties == null)
-                    {
-                        log.Verbose("Failed to find deployment.Properties");
-                        return;
-                    }
-
-                    log.Verbose($"Provisioning state: {deployment.Properties.ProvisioningState}");
-                    switch (deployment.Properties.ProvisioningState)
-                    {
-                        case "Succeeded":
-                            log.Info($"Deployment {deploymentName} complete.");
-                            log.Info(GetOperationResults(armClient, resourceGroupName, deploymentName));
-                            CaptureOutputs(deployment.Properties.Outputs?.ToString(), variables);
-                            continueToPoll = false;
-                            break;
-
-                        case "Failed":
-                            throw new CommandException($"Azure Resource Group deployment {deploymentName} failed:\n" + GetOperationResults(armClient, resourceGroupName, deploymentName));
-
-                        case "Canceled":
-                            throw new CommandException($"Azure Resource Group deployment {deploymentName} was canceled:\n" + GetOperationResults(armClient, resourceGroupName, deploymentName));
-
-                        default:
-                            if (currentPollWait < maxWaitSeconds)
-                            {
-                                var temp = previousPollWait;
-                                previousPollWait = currentPollWait;
-                                currentPollWait = temp + currentPollWait;
-                            }
-
-                            break;
-                    }
-                }
+                var deploymentResult = await AsyncResourceGroupPollingTimeoutPolicy.ExecuteAsync(async timeoutCancellationToken =>
+                                                                                                 {
+                                                                                                     var delayStrategy = DelayStrategy.CreateExponentialDelayStrategy(TimeSpan.FromSeconds(1), PollingTimeout);
+                                                                                                     var result = await deploymentOperation.WaitForCompletionAsync(delayStrategy, timeoutCancellationToken);
+                                                                                                     return result;
+                                                                                                 },
+                                                                                                 CancellationToken.None);
+                log.Info($"Deployment completed with status: {deploymentResult.Data.Properties?.ProvisioningState}");
             }
+            catch
+            {
+                log.Error("Error polling for deployment completion");
+                throw;
```

Comment:
How does this exception get handled? Does it need to be wrapped in a `CommandException` for the correct exit status?

Reply:
It bubbles up and hits this [exception handler](https://github.com/OctopusDeploy/Calamari/blob/a5aeca4690dfc5dbde4494fe7c30cb8312c7e1a8/source/Calamari.Common/CalamariFlavourProgram.cs#L82-L85) which logs out the exception and returns exit code 100. I could wrap it in a `CommandException` so it returns exit code 1 but it might potentially lose some detail in the error log

![Screenshot 2024-04-12 at 11 13 20 am](https://github.com/OctopusDeploy/Calamari/assets/3272176/e06c9c74-2a94-4196-980d-14fdbc8dce73)

Reply:
I think you can pass the actual exception into `CommandException` as the innerException and you would still get the full stack trace of the inner exception. I'm not to concerned either way because I don't know the fine distinction between a command error vs other error exit status code.